crates/iota-grpc-api/src/types.rs

```diff
@@ -167,11 +190,176 @@ where
     }
 }
 
+/// Wrapper around RestStateReader that implements the GrpcStateReader trait
+#[derive(Clone)]
+pub struct GrpcReader {
+    state_reader: Arc<dyn GrpcStateReader>,
+}
+
+impl GrpcReader {
+    pub fn new(state_reader: Arc<dyn GrpcStateReader>) -> Self {
+        Self { state_reader }
+    }
+
+    pub fn from_rest_state_reader(state_reader: Arc<dyn RestStateReader>) -> Self {
+        Self {
+            state_reader: Arc::new(RestStateReaderAdapter {
+                inner: state_reader,
+            }),
+        }
+    }
+
+    pub fn get_epoch_last_checkpoint(
+        &self,
+        epoch: u64,
+    ) -> Result<Option<CertifiedCheckpointSummary>, Box<dyn std::error::Error + Send + Sync>> {
+        self.state_reader.get_epoch_last_checkpoint(epoch)
+    }
+}
+
+/// Adapter that implements GrpcStateReader for RestStateReader
+pub struct RestStateReaderAdapter {
+    inner: Arc<dyn RestStateReader>,
+}
+
+impl GrpcStateReader for RestStateReaderAdapter {
+    fn get_latest_checkpoint_sequence(&self) -> Option<u64> {
+        Some(*self.inner.get_latest_checkpoint().sequence_number())
+    }
+
+    fn get_checkpoint_summary(&self, seq: u64) -> Option<CertifiedCheckpointSummary> {
+        self.inner
+            .get_checkpoint_by_sequence_number(seq)
+            .map(CertifiedCheckpointSummary::from)
+    }
+
+    fn get_checkpoint_data(&self, seq: u64) -> Option<CheckpointData> {
+        let summary = self.inner.get_checkpoint_by_sequence_number(seq)?;
+        let contents = self.inner.get_checkpoint_contents_by_sequence_number(seq)?;
+        Some(self.inner.get_checkpoint_data(summary, contents))
+    }
+
+    fn get_epoch_last_checkpoint(
+        &self,
+        epoch: u64,
+    ) -> Result<Option<CertifiedCheckpointSummary>, Box<dyn std::error::Error + Send + Sync>> {
+        match self.inner.get_epoch_last_checkpoint(epoch) {
+            Ok(Some(checkpoint)) => Ok(Some(CertifiedCheckpointSummary::from(checkpoint))),
+            Ok(None) => Ok(None),
+            Err(e) => Err(Box::new(e) as Box<dyn std::error::Error + Send + Sync>),
+        }
+    }
+}
+
 #[derive(Clone)]
 pub struct Reader {
     pub state_reader: Arc<dyn RestStateReader>,
 }
 
+impl GrpcReader {
+    fn get_full_checkpoint_data(&self, seq: u64) -> Option<CheckpointData> {
+        self.state_reader.get_checkpoint_data(seq)
+    }
+
+    pub fn get_latest_checkpoint_sequence(&self) -> Option<u64> {
+        self.state_reader.get_latest_checkpoint_sequence()
+    }
+
+    pub fn create_checkpoint_stream<T>(
```

Comment:
This function looks weird as a method of GrpcReader. It looks like it should be a standalone function or an extension method of CheckpointReader trait.